### PR TITLE
Improve type annotations for `execute_values`.

### PR DESCRIPTION
**Pull request by reivilibre**
Should be a commit-by-commit review.

This was something that I agreed to look into whilst I was in the area.
Turns out the `*args` was not necessary; there's only one arg here (an iterable of value rows).

I also added a comment to `_do_execute`, where we could improve the type annotations easily a little bit
once `ParamSpec` is more usable.

### CODE REVIEW

File: synapse/storage/database.py [hunk 1]
```diff
@@ -326,6 +324,16 @@ def _make_sql_one_line(self, sql: str) -> str:
...
+    # TODO(ParamSpec):
+    #     Once Mypy supports Concatenate, this could be written as
+    #     def _do_execute(
+    #         self,
+    #         func: Callable[Concatenate[str, P], R],
+    #         sql: str,
+    #         *args: P.args,
+    #         **kwargs: P.kwargs,
+    #     ) -> R:
+    #     so long as we also pass kwargs.
```
Comment (clokep) [inline synapse/storage/database.py:336]: Seems better as an issue that we file?

File: synapse/storage/database.py [hunk 2]
```diff
@@ -305,15 +307,11 @@ def execute_values(self, sql: str, *args: Any, fetch: bool = True) -> List[Tuple
...
         return self._do_execute(
-            # Type ignore: mypy is unhappy because if `x` is a 5-tuple, then there will
-            # be two values for `fetch`: one given positionally, and another given
-            # as a keyword argument. We might be able to fix this by
-            # - propagating the signature of psycopg2.extras.execute_values to this
-            #   function, or
-            # - changing `*args: Any` to `values: T` for some appropriate T.
-            lambda *x: execute_values(self.txn, *x, fetch=fetch),  # type: ignore[misc]
+            lambda the_sql, the_values: execute_values(
+                self.txn, the_sql, the_values, fetch=fetch
```
Comment (clokep) [inline synapse/storage/database.py:311]: (This code was pre-existing, but...)

`the_sql` doesn't change for each iterate, correct? So really no reason to pass it in for each one?

Comment (reivilibre) [reply synapse/storage/database.py:311]: Sorry, what do you mean?

`the_sql` is given by `_do_execute`.

`the_values` are just a pass-through of `values` and could have been part of the lambda's closure, but in analogue to `execute_batch` (and indeed the way this code was written before), it's passed through. If we want to change this I suppose I have no objections, though I wonder if the idea was that `_do_execute` could have transformed the args if necessary...?

File: synapse/storage/database.py (hunk 1 above)
Comment (reivilibre) [reply synapse/storage/database.py:336]: yeah, if you think so, could do that instead.

File: synapse/storage/database.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/database.py:311]: Errr, yes...that didn't make sense.

My confusion is -- why are we passing in `fetch` directly into the lambda, but we pass in `the_sql` and `the_values` as arguments to the lambda?

**Review by clokep — APPROVED** [review]
Seems fine, there's a pending question but won't block merging this on it!

File: synapse/storage/database.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/database.py:311]: Ah, I did not realize that `_do_execute` modified the passed in `sql`!

Comment (reivilibre) [reply synapse/storage/database.py]: I've changed this around a bit, is that better to you?

Comment (clokep) [reply synapse/storage/database.py]: I do find that clearer, yes!

**Pull request merged**